feat: replace temp file for vvm error

s3bc40 opened the pull request:
Related issue: #170 

Address temporary path issues in the VVM compiler. 

WIP since the problem could be in other CLI commands. Proposal to accept @PatrickAlphaC. Maybe we need to find a way to make it global and avoid duplication.

### Error handling enhancements:

* Added handling for `VVMVyperError` in the `compile` function. This includes a workaround to replace temporary paths in the error message with the actual contract path, improving the clarity of error logs.

### Import updates:

* Added `re` module import to facilitate regular expression operations for the error-handling logic.
* Imported `VyperError` as `VVMVyperError` from `vvm.exceptions` to handle VVM-specific errors.

s3bc40 commented:
I'll try to add a global handler error for VVM to avoid writing the same stuff in all cli commands: 

https://dnmtechs.com/python-3-implementing-a-global-error-handler-for-any-exception/

s3bc40 commented:
Well, I think I can not make it global since I need the `contract` path to change the `stderr` message. And for now I can only do it on the `compile` cli. I tried to do it on `run` but it is a Python script working with the `VyperContract` deployed.

I have refactored to avoid code to be duplicated, but for now it is only available in the compile cli.

PatrickAlphaC commented:
Big thanks for working on this, and I LOVE the simplicity. Can you show me an example of this in action? Like, maybe give me a minimal reproducible example so I can test the functionality works as you expect? 

A test might be good too, but if you can just give me the code you expect to behave a certain way

PatrickAlphaC reviewed (changes requested):
see comment

s3bc40 commented:
Yes I can do that, sorry I should have thought of making a test for this rather than changing the `Counter.vy` inside the `data/installation_project` and compile it locally with `mox compile`

```diff
# SPDX-License-Identifier: MIT
- # pragma version ^0.4.1
+ # pragma version 0.4.0

number: public(uint256)

event Test:
    name: String[32]
    number: uint256

@external
def set_number(new_number: uint256):
+    log Test("Hello") # IT SHOULD BRING THE VVM ERROR ON COMPILE
    self.number = new_number

@external
def increment():
    self.number += 1
```

I think I'll use the `test_project` folder and add new contracts with `0.4.0` and an error inside. And then add test cases for it.

s3bc40 commented:
Well, I was going for a quick addition to the test suit, but I ended up crashing on a few problems on the way, sorry. 

Essentially, from my commit message:

- Added VVM Vyper error cases to the test suite.
- Fixed the install command to always create lib directories by default if they don't exist.
- Fix general issues on unit tests and integration tests (version mismatch, assertions, etc.).

Few VVM Vyper error cases are now covered.

---

Before

```
Error compiling: /tmp/vyper-jo_5g4h4.vy

> stderr:
vyper.exceptions.StructureException: Expressions without assignment are disallowed

  (hint: did you mean to assign the result to a variable?)

  contract "/tmp/vyper-jo_5g4h4.vy:7", function "set_number", line 7:4
       6 def set_number(new_number: uint256):
  ---> 7     Hello
  -----------^
       8     self.number = new_number
```

After
```

Error compiling: /tmp/vyper-e79dmdv1.vy

> stderr:
vyper.exceptions.StructureException: Expressions without assignment are disallowed

  (hint: did you mean to assign the result to a variable?)

  contract "/home/s3bc40/gh-projects/moccasin/tests/data/vvm_error_project/src/AssignError.vy:7", function "set_number", line 7:4
       6 def set_number(new_number: uint256):
  ---> 7     Hello
  -----------^
       8     self.number = new_number
```

---

@PatrickAlphaC tell me if you want some things back to where they were like the installation part. What I have added is what I wanted to add as a feature at first, but maybe that's not what you wanted.

PatrickAlphaC commented:
hmmm... I can't replicate what you're getting:


Your code
```console
% mox build AssignError.vy
Running compile command...
Traceback (most recent call last):
  File "/PATH/code/moccasin/moccasin/commands/compile.py", line 172, in compile_
    deployer: VyperDeployer | VVMDeployer = load_partial(
                                            ~~~~~~~~~~~~^
        str(contract_path), compiler_args
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/boa/interpret.py", line 287, in load_partial
    return loads_partial(
        f.read(), name=filename, filename=filename, compiler_args=compiler_args
    )
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/boa/interpret.py", line 276, in loads_partial
    return _loads_partial_vvm(source_code, version, name, filename)
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/boa/interpret.py", line 329, in _loads_partial_vvm
    ret = _disk_cache.caching_lookup(cache_key, _compile)
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/boa/util/disk_cache.py", line 75, in caching_lookup
    res = func()
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/boa/interpret.py", line 308, in _compile
    return vvm.compile_source(
           ~~~~~~~~~~~~~~~~~~^
        source_code, vyper_version=version, base_path=base_path
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/vvm/main.py", line 69, in compile_source
    compiler_data = _compile(
        vyper_binary=vyper_binary,
    ...<4 lines>...
        output_format=output_format,
    )
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/vvm/main.py", line 147, in _compile
    stdoutdata, stderrdata, command, proc = wrapper.vyper_wrapper(
                                            ~~~~~~~~~~~~~~~~~~~~~^
        vyper_binary=vyper_binary, f=output_format, p=base_path, **kwargs
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/vvm/wrapper.py", line 137, in vyper_wrapper
    raise VyperError(
    ...<5 lines>...
    )
vvm.exceptions.VyperError: An error occurred during execution
> command: `/PATH/.vvm/vyper-0.4.0 /var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy -f combined_json -p .`
> return code: `1`
> stdout:
Error compiling: /var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy

> stderr:
vyper.exceptions.StructureException: Expressions without assignment are disallowed

  (hint: did you mean to assign the result to a variable?)

  contract "/var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy:7", function "set_number", line 7:4 
       6 def set_number(new_number: uint256):
  ---> 7     Hello
  -----------^
       8     self.number = new_number

[95413] Failed to execute script 'vyper_compile' due to unhandled exception!

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/PATH/code/moccasin/.venv/bin/mox", line 10, in <module>
    sys.exit(main())
             ~~~~^^
  File "/PATH/code/moccasin/moccasin/__init__.py", line 8, in main
    __main__.main(sys.argv[1:])
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^
  File "/PATH/code/moccasin/moccasin/__main__.py", line 58, in main
    import_module(f"moccasin.commands.{command_to_run}").main(args)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/PATH/code/moccasin/moccasin/commands/compile.py", line 45, in main
    compile_(
    ~~~~~~~~^
        contract_path,
        ^^^^^^^^^^^^^^
    ...<2 lines>...
        write_data=True,
        ^^^^^^^^^^^^^^^^
    )
    ^
  File "/PATH/code/moccasin/moccasin/commands/compile.py", line 189, in compile_
    handle_vvm_error(vvm_exc, contract_path)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/PATH/code/moccasin/moccasin/_error_utils.py", line 21, in handle_vvm_error
    raise VVMVyperError(
    ...<4 lines>...
    )
vvm.exceptions.VyperError: An error occurred during execution
> command: `/PATH/.vvm/vyper-0.4.0 /var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy -f combined_json -p .`
> return code: `1`
> stdout:
Error compiling: /var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy

> stderr:
vyper.exceptions.StructureException: Expressions without assignment are disallowed

  (hint: did you mean to assign the result to a variable?)

  contract "/var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy:7", function "set_number", line 7:4 
       6 def set_number(new_number: uint256):
  ---> 7     Hello
  -----------^
       8     self.number = new_number

[95413] Failed to execute script 'vyper_compile' due to unhandled exception!
```

The current code:

```
% mox build AssignError.vy
Running compile command...
Traceback (most recent call last):
  File "/PATH/code/moccasin/moccasin/commands/compile.py", line 172, in compile_
    deployer: VyperDeployer | VVMDeployer = load_partial(
                                            ~~~~~~~~~~~~^
        str(contract_path), compiler_args
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/boa/interpret.py", line 287, in load_partial
    return loads_partial(
        f.read(), name=filename, filename=filename, compiler_args=compiler_args
    )
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/boa/interpret.py", line 276, in loads_partial
    return _loads_partial_vvm(source_code, version, name, filename)
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/boa/interpret.py", line 329, in _loads_partial_vvm
    ret = _disk_cache.caching_lookup(cache_key, _compile)
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/boa/util/disk_cache.py", line 75, in caching_lookup
    res = func()
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/boa/interpret.py", line 308, in _compile
    return vvm.compile_source(
           ~~~~~~~~~~~~~~~~~~^
        source_code, vyper_version=version, base_path=base_path
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/vvm/main.py", line 69, in compile_source
    compiler_data = _compile(
        vyper_binary=vyper_binary,
    ...<4 lines>...
        output_format=output_format,
    )
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/vvm/main.py", line 147, in _compile
    stdoutdata, stderrdata, command, proc = wrapper.vyper_wrapper(
                                            ~~~~~~~~~~~~~~~~~~~~~^
        vyper_binary=vyper_binary, f=output_format, p=base_path, **kwargs
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/PATH/code/moccasin/.venv/lib/python3.13/site-packages/vvm/wrapper.py", line 137, in vyper_wrapper
    raise VyperError(
    ...<5 lines>...
    )
vvm.exceptions.VyperError: An error occurred during execution
> command: `/PATH/.vvm/vyper-0.4.0 /var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy -f combined_json -p .`
> return code: `1`
> stdout:
Error compiling: /var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy

> stderr:
vyper.exceptions.StructureException: Expressions without assignment are disallowed

  (hint: did you mean to assign the result to a variable?)

  contract "/var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy:7", function "set_number", line 7:4 
       6 def set_number(new_number: uint256):
  ---> 7     Hello
  -----------^
       8     self.number = new_number

[95413] Failed to execute script 'vyper_compile' due to unhandled exception!

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/PATH/code/moccasin/.venv/bin/mox", line 10, in <module>
    sys.exit(main())
             ~~~~^^
  File "/PATH/code/moccasin/moccasin/__init__.py", line 8, in main
    __main__.main(sys.argv[1:])
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^
  File "/PATH/code/moccasin/moccasin/__main__.py", line 58, in main
    import_module(f"moccasin.commands.{command_to_run}").main(args)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/PATH/code/moccasin/moccasin/commands/compile.py", line 45, in main
    compile_(
    ~~~~~~~~^
        contract_path,
        ^^^^^^^^^^^^^^
    ...<2 lines>...
        write_data=True,
        ^^^^^^^^^^^^^^^^
    )
    ^
  File "/PATH/code/moccasin/moccasin/commands/compile.py", line 189, in compile_
    handle_vvm_error(vvm_exc, contract_path)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/PATH/code/moccasin/moccasin/_error_utils.py", line 21, in handle_vvm_error
    raise VVMVyperError(
    ...<4 lines>...
    )
vvm.exceptions.VyperError: An error occurred during execution
> command: `/PATH/.vvm/vyper-0.4.0 /var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy -f combined_json -p .`
> return code: `1`
> stdout:
Error compiling: /var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy

> stderr:
vyper.exceptions.StructureException: Expressions without assignment are disallowed

  (hint: did you mean to assign the result to a variable?)

  contract "/var/folders/h6/j1dkww4j5k9bvjpqtkyqcp3r0000gn/T/vyper-8oqdwdvx.vy:7", function "set_number", line 7:4 
       6 def set_number(new_number: uint256):
  ---> 7     Hello
  -----------^
       8     self.number = new_number

[95413] Failed to execute script 'vyper_compile' due to unhandled exception!
```

PatrickAlphaC commented:
Looks like it might not always be `tmp` that is being targeted?

s3bc40 commented:
Ah maybe it's linked to MacOS:
```
On MacOS, the main temporary folder can be accessed by running echo $TMPDIR in the Terminal, which usually returns a path like /var/folders/s9/yl377tk95k9353s3k1jpn4_80000gn/T/. There is also a /tmp directory that is universal to all users and is cleared after a reboot.
```

I might take everything into account

s3bc40 commented:
```py
def test_compile_with_vvm_assign_error(
    mox_path, vvm_error_cleanup_out_folder, vvm_error_cleanup_dependencies_folder
):
    current_dir = Path.cwd()
    try:
        os.chdir(current_dir.joinpath(VVM_ERROR_PROJECT_PATH))
        subprocess.run(
            [mox_path, "build", "AssignError.vy"],
            check=True,
            capture_output=False, # -->  FACEPALM
            text=True,
        )
```
--- 

Other than this, I reworked the feature to take into the differences in `temp_dir` path, with the help of AI + review/rework:

```py
    original_stderr = error.stderr_data
    # @dev tempfile.gettempdir() retrieves the system's standard temporary directory path.
    # @dev re.escape() to ensure that any special characters in the temporary directory path are treated literally in the regular expression.
    temporary_path_pattern = (
        re.escape(tempfile.gettempdir()) + r"[/\\]vyper-[a-zA-Z0-9_-]+\.vy:(\d+)"
    )
    replacement_string = f"{str(contract_path)}:\\1"
```

And I added a unit test specific to this feature, without the compile stuff, to ensure that the regex is working.

I hope it will fix everything :crossed_fingers:

PatrickAlphaC commented:
I'm getting nervous this is going to be a tricky feature to keep maintained... wondering if we can put it on ice for a bit. Would way rather the Metamask feature <3

Or a trezor/aws/ledger integration!!!

s3bc40 commented:
Well, the better solution would be to have this feature inside the VVM itself so that it always return the actual path. But it might not be a priority for the team.

It might be better to focus on the other subjects :+1:

PatrickAlphaC commented:
Going to close for now, feel free to reopen in the future
